### ISSUE TITLE
[indent] conflicts with TypeScripts auto-formatting for a single semicolon

### ISSUE THREAD

**Issue by AlCalzone** [typescript-eslint/typescript-eslint#424]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/indent": ["error", "tab"],
  }
}
```

```TS
declare const bar: any;
function foo(): void {
	bar
		.baz()
		.bap()
		;
}
```
^ This is how it is formatted after applying auto-formatting in VSCode.

**Expected Result**
no error

**Actual Result**
![grafik](https://user-images.githubusercontent.com/17641229/55881131-604d0280-5ba2-11e9-9409-7ac4ede75bc1.png)

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.6.0` |
| `@typescript-eslint/parser`        | `1.6.0` |
| `TypeScript`                       | `3.4.2` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `10.15.2` |
| `npm`                              | `6.4.1` |

**Comment by bradzacher** [comment]
what was the code before fixing?
It's a bit hard to debug the exact case with only the fixed code.

**Comment by AlCalzone** [comment]
@bradzacher It doesn't really matter how you indent it, the above code is the output after you run the `Format document` command in VSCode or save the file with the `editor.formatOnSave` option set to true. AFAIK that uses the default typescript formatter. I don't have any special formatting extensions.

ESLint wants 1 tab indentation for the semicolon, VSCode/TypeScript wants 2 tabs. When I run the ESLint fixer, it becomes 1 tab. When I change something and save, it becomes 2 tabs again.

**Comment by bradzacher** [comment]
It is not our intention to be compatible with the vscode formatter - in fact we aren't actually a formatter at all.

If you want to use prettier to format your code, or the typescript language service to format your code, then it is up to you to configure ESLint to match your desired formatter.

I would suggest disabling rules like `indent` as they will never match the implementation of the typescript formatter.
As an example, [prettier itself provides a predefined configuration](https://github.com/prettier/eslint-config-prettier) to disable eslint's formatting rules so that they don't conflict with prettier's formatting style.

**Comment by AlCalzone** [comment]
Hmm thats disappointing. The [roadmap for migration](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/ROADMAP.md#maintainability) lists the `@typescript-eslint/indent` rule as the (complete) replacement for TSLints `indent` rule.

Seems I'll have to include yet another package (plus VSCode plugin) just to get rid of conflicting opinions about semicolon indentation.

**Comment by bradzacher** [comment]
The indent rule is, for all intents and purposes, a complete replacement. Though there are probably going to be minute differences between the two because there are hundreds if not thousands of possible cases.

Formatting is a difficult and complex business. It's really hard to get it perfect so that everyone agrees. ESLint's rules all taken together make one formatting style, prettier is another, and the typescript language service provides a third. I'm sure there are countless other formatters out there. There's no way for us to support and maintain compatibility for every single style, so we don't support any.

I would suggest if you can to use prettier. You won't find a better javascript/typescript formatter IMO. It's opinionated and very strict - you know that everything will be formatted the same regardless of how many times its run or how someone actually writes the code. They also have an [eslint plugin](https://github.com/prettier/eslint-plugin-prettier) which will help enforce that people have run prettier on their code.

---

To be fair as well, this is actually the first time I've seen someone put the semicolon on a new line because that "goes against" ASI (automatic semicolon insertion).
i.e. unless I'm mistaken, your code with ASI applied will be equivalent to:
```TS
declare const bar: any;
function foo(): void {
	bar
		.baz()
		.bap();
		;
}
```

So ESLint is parsing the semicolon as not part of the same statement, because technically it's not, hence it expects 1 tab instead of 2.

**Comment by AlCalzone** [comment]
My assumption was (now that the TypeScript ecosystem is moving to ESLint) that at least the "new" rules prefixed with `@typescript-eslint` would at least agree with the language service by default.

> I would suggest if you can to use prettier. [...] They also have an eslint plugin which will help enforce that people have run prettier on their code.

Thats what I did now. I do not agree with all of its formatting choices, but that is by design.

Regarding the lone semicolon:
This is one of the things (alongside trailing commas, leading `|` in multiline union types, etc...) I used to do to limit git diffs to the lines that actually changed. I never thought about ASI here. For example, if I add another method to the chain, the `git diff` looks like
```diff
  declare const bar: any;
  function foo(): void {
  	bar
  		.baz()
  		.bap()
+ 		.anotherMethod()
  		;
  }
```
which is much clearer than
```diff
  declare const bar: any;
  function foo(): void {
  	bar
  		.baz()
-   		.bap();
+ 		.bap()
+ 		.anotherMethod();
  }
```
